### PR TITLE
[MM-26589] Add context version to all public methods

### PR DESCRIPTION
**Pull request by ethervoid**
#### Summary

To be able to cancel the request by timeout or deadline we need to pass the context to the request.
    
This PR a new *WithContext method to all the exposed methods so the user can:

- Use the current method, which is going to create a [context.Background](https://golang.org/pkg/context/#Background) context
- Use the *WithContext method and pass their own context

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-26589

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,11 +1,11 @@
 version: 2
 jobs:
   build:
-    working_directory: /go/src/github.com/jszwedko/go-circleci
+    working_directory: /go/src/github.com/mattermost/go-circleci
     environment:
       TEST_RESULTS: /tmp/test-results
     docker:
-    - image: circleci/golang:1.10
+    - image: circleci/golang:1.13
```
Comment (agnivade) [inline .circleci/config.yml:8]: Any reason we can't go for 1.14?

File: circleci_test.go [hunk 2]
```diff
@@ -101,12 +104,28 @@ func TestClient_request(t *testing.T) {
...
+	defer teardown()
+	client.Token = "ABCD"
+	mux.HandleFunc("/me", func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(1 * time.Second)
+	})
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Microsecond)
+	defer cancel()
+	err := client.request(ctx, "GET", "/me", &User{}, nil, nil)
+	if err == nil || !errors.As(err, &context.DeadlineExceeded) {
```
Comment (agnivade) [inline circleci_test.go]: Sorry, I was a bit wrong. `context.DeadlineExceeded` is a variable. So `if err == nil || !errors.Is(err, context.DeadlineExceeded) {` is more suited here.

Sorry for the double-work.

**Review by agnivade — COMMENTED** [review]
Just a last round of comments.

File: .circleci/config.yml (hunk 1 above)
Comment (ethervoid) [reply .circleci/config.yml:8]: Just to be conservative by now and use the minimum required version

File: circleci_test.go (hunk 2 above)
Comment (ethervoid) [reply circleci_test.go]: Done, no problem. Also, I'm learning hehehe

File: .circleci/config.yml (hunk 1 above)
Comment (agnivade) [reply .circleci/config.yml:8]: The minimum required version is actually 1.14 for server. Which minimum required version are you referring to?

Comment (ethervoid) [reply .circleci/config.yml:8]: To the use of `RequestWithContext` given that this project could be used by other projects that are not tied to 1.14 but I can change to that version without problem

Comment (agnivade) [reply .circleci/config.yml:8]: Ah no worries. This is fine for now. Thanks.

**Review by agnivade — APPROVED** [review]

**Pull request merged**